src/components/files/FilePreview/CsvPreview.tsx

```diff
@@ -0,0 +1,50 @@
+import { Grid } from '@material-ui/core';
+import Papa, { ParseResult } from 'papaparse';
+import React, { FC, useCallback, useEffect, useState } from 'react';
+import { useFileActions, usePreviewError } from '../FileActions';
+import { PreviewerProps } from './FilePreview';
+import { PreviewLoading } from './PreviewLoading';
+import { ColumnData, SpreadsheetView } from './SpreadsheetView';
+
+export const CsvPreview: FC<PreviewerProps> = ({ downloadUrl }) => {
+  const [csvData, setCsvData] = useState<ParseResult<string[]>['data']>([]);
+  const { previewLoading, setPreviewLoading } = useFileActions();
+  const handleError = usePreviewError();
+
+  // ignoring result.errors for now and just using result.data
+  const handleReadComplete = useCallback(
+    ({ data }) => {
+      setCsvData(data);
+      setPreviewLoading(false);
+    },
+    [setPreviewLoading]
+  );
+
+  useEffect(() => {
+    setPreviewLoading(true);
+    Papa.parse(downloadUrl, {
+      complete: handleReadComplete,
+      download: true,
+      error: () => handleError('Could not read CSV'),
+    });
+  }, [downloadUrl, setPreviewLoading, handleReadComplete, handleError]);
```

Comment:
So currently for most (all?) preview parsing we have 3 callbacks for loading, error, and data.
It better to use an async function that either returns data, throws an error, and after either loading is set to false. This matches the same pattern that final form submit handler uses.

Maybe this "parser" function is passed to separately and uses a generic so the state can be passed in instead of having to be managed in every previewer component.

Reply:
Okay, I didn't do exactly this because I was able to make all the previewer components accept a `File` instead of a URL. This let us centralize the file-retrieval logic but leave each individual previewer in charge of parsing and setting the loading/error state at the end of parsing using the same values/functions from `FileActionsContext`. So it's more DRY, but there's no real passing around of state.